dig.go

```diff
@@ -73,26 +72,28 @@ type Container struct {
 	// index map[reflect.Type]key
 }
 
-// New constructs a ready-to-use Container.
+// New constructs a Container.
 func New(opts ...Option) *Container {
 	return &Container{
 		nodes: make(map[key]*node),
 		cache: make(map[key]reflect.Value),
 	}
 }
 
-// Provide teaches the Container how to construct one or more new types.
+// Provide teaches the container how to build values of one or more types.
 //
-// Any function passed to Provide is assumed to be a constructor. Constructors
-// can take any number of parameters, which will be supplied by the Container
-// on demand. They must return at least one non-error value, all of which are
-// then available in the Container. If the last returned value is an error, the
-// Container inspects it to determine whether the constructor succeeded or
-// failed. Regardless of position, returned errors are never put into the
-// Container's dependency graph.
+// The first argument of Provide is a function which accepts zero or more
+// parameters and returns one or more results. The function may optionally
+// return an error to indicate that it failed to build the value. This
+// function will be treated as the constructor for all the types it returns.
+// This function will be called AT MOST ONCE when a type produced by it, or a
```

Comment:
Isn't it going to be executed `exactly ONCE` because it is requested by Invoke?
Should we also add explicitly that this function will not be called if returned types are not requested?